tests/base/task_11.cc

```diff
@@ -0,0 +1,99 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2009 - 2015 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+// test calling MultithreadInfo::set_thread_limit() multiple times and the
+// effect for running tasks
+
+#include "../tests.h"
+#include <iomanip>
+#include <fstream>
+#include <unistd.h>
+
+#include <deal.II/base/thread_management.h>
+
+Threads::Mutex mutex;
+unsigned int n_running;
+unsigned int n_max_running;
+
+void a_task ()
+{
+  mutex.acquire();
+  n_running++;
+  n_max_running=std::max(n_max_running, n_running);
+  mutex.release();
+
+  sleep (1);
+
+  mutex.acquire();
+  n_running--;
+  mutex.release();  
+}
+
+
+void test()
+{
+  Threads::TaskGroup<> tg;
+
+  n_running = 0;
+  n_max_running = 0;
+
+  // force all tasks to wait until we are done starting
+  mutex.acquire();
+
+  for (unsigned int t=0;t<10;++t)
+    tg += Threads::new_task(a_task);
+
+  mutex.release();
+  
+  tg.join_all();
+  deallog << "max concurrent running: " << n_max_running
+	  << " should be: " << MultithreadInfo::n_threads()
```

Comment:
@bangerth How can this work? If the <code>join_all</code> comes before the <code>mutex.release</code> the test deadlocks because no task will ever finish.